### ISSUE TITLE
Example of a species habitat analysis and possible BAP.

### ISSUE THREAD

**Issue by nmtarr** [usgs-bcb/bcb-dm#43]
The Core Principles section of the “read me” file for next week’s meeting mentions analyses of habitat data and projects in the RWO’s.  There are a few analyses in the NCSU RWO with similar methods that I envisioned as candidates for BAPs.  I’m not sure exactly what a BAP will look like or if my understanding of them is correct, but I thought I’d go ahead and share an example of an analysis because it sounds like we are approaching discussions analyses.  I realize this may be tangential to what we actually work on and therefore unnecessary.

This example is an analysis of the importance of woody wetlands of the Southeast for wildlife.  The other analyses in our RWO with similar methods are related to drought and forestry in the Southeast, although Matt and I are covering the forestry topic in our collaboration with folks from the University of Utrecht.  I wrote up a [short report](https://docs.google.com/document/d/1SscvHR2ayVaHhKQ4DaCRW7b2wsfw_hNs8QPa1IkJXpQ/edit?usp=sharing) on this woody wetland analysis.  The entire analysis is scripted in python as [top-level scripts](https://github.com/nmtarr/SEWoodyWetlandAnalysis) that use GAP packages available [here ](https://github.com/nmtarr/GAPAnalysis)and [here](https://github.com/nmtarr/GAPProduction).  If you look this stuff over, please keep the following things in mind:
- This is just a draft that could be lengthened and changed as necessary.  I was planning on making it more complete before sharing, and I expect it to eventually change per comments.
- You’ll notice some sloppiness with language.  Eventually, the collection of ecological systems I focused on would be called either floodplain forests or SE woody wetlands or something else.  This draft starts out calling them woody wetlands and then goes to floodplain forests.
- The analysis uses a threshold of 50% for determining “floodplain species”.  Maybe a BAP version of this analysis could allow the user to pick the threshold.
- The Discussion is underdeveloped, especially with regards to answering “so what?” and in interpreting/discussing the species on the list.  Could the TIR automatically provide access to some information regarding threats and population status of each species?
- The formatting of figures is rough because I didn’t want to fine-tune it until I knew the format of the report or BAP.
- Other results are possible, such as finding out what proportion of CONUS and SE hotspots are floodplain forests, the rank (percentile) of each floodplain species protection in status 1 or 2 lands, and who owns the protected floodplain forests.
- The paper, especially Methods, is written with you (BCB) as the audience, not as they would appear in the NBM or in a manuscript.
- The slowest process within this analysis was the overlay of habitat maps with the floodplain forest map, which ran for almost a week on four kernels.  The overlay process will be necessary for the other analyses that will use these methods (e.g. assessment of habitat in drought zones), but this analysis (SE woody wetlands) could have actually just used information from the Analytic Database since the area of interest is derived from the 2001 GAP Land Cover Map.
- As the analysis is scripted and we should soon have more current maps (2011), I would hope to base a final version on the more current habitat maps.

**Comment by skybristol** [comment]
Fantastic! This is exactly what I want to get into over the next couple of days. One of the things we are working to do is get our infrastructure to the point where this type of analysis can be done fully with online data and APIs instead of requiring the specialized setup of backend dependencies. That capacity will allow analysis packages to be fully traceable through their full provenance chain and publishable online as executable code. It will also set us up to be able to address the issue you raised of the weeklong process to run a computationally (and memory, I assume) intense process where we can move the online data and processing code to the most appropriate batch of resources to operate efficiently.

One of the interesting new ideas that your discussion here and the coded analysis you put together sparks for me is in the "so what" part of this. It makes me think that we might also want to create something like a "Habitat Information Registry" construct in the Biogeographic Information System, based partly on Ecological Systems (NVCS foundation?) and then additional work on classifying freshwater habitats, estuarine and benthic habitats, etc. For our mission within DOI, we're ultimately working to inform management strategies and priorities with our work. Much of that for us comes down to not only pointing out those areas that are rare or threatened and in need of conservation measures but what the potential management actions might be that would most effectively address the threats. Just like we're assembling a body of literature and other information along with structured attributes on species, we could systematically assemble further information on habitat management methods to compliment the systematic classification information in the NVCS. With that kind of thing in place, we can then pair analyses on potential future threats with possible response strategies.

I did some work in the last few days on rearchitecting the TIR with some new and simpler technologies. This [notebook](https://github.com/usgs-bcb/gap_species/blob/master/GAP%20Species%20summary%20from%20new%20version%20of%20TIR.ipynb) runs through a quick summary of the different processors I've built so far and the information they are returning. I hope to get into refining the existing processes together as a group, and we need to focus those discussions on analytical processes that we want to follow and how best to tee up the information to support those.

**Comment by nmtarr** [comment]
I definitely think moving the analysis online would be ideal.  Am I right that the biggest challenge with that task will be how to speed up the process of overlaying our giant rasters? That process is always SUPER slow.  If so, we might want to focus on that challenge up front.  That's a process that frequently comes into play in various analyses and other tasks, so there would be a big payoff if cracking that nut.

I think your idea about a habitat information registry is a logical place to go.  I also suspect that what you are thinking of would/could overlap with a "wildlife-centric classification" that I began developing to simplify the process of building habitat models.  It also has use for analyses (it's used in the one I shared) and I think it could offer a solution to the problem of evolving habitat classifications or linking models to alternate habitat classifications, although that would require enormous amounts of work.  I'm happy to elaborate if we get into this topic.  What I've got so far is stored and accessible with [this function](https://github.com/nmtarr/GAPProduction/blob/master/gapproduction/wildclass.py).

**Comment by McKerrow** [comment]
We might want to take a look at some standard definitions related to riparian, floodplain etc.  to see if we can easily link to other systems in what we call out as floodplain.  I just had a meeting with NRCS folks who are using the gridded ssurgo & our land cover to see if they can inform each other - that might be something to take a look at as well.
https://www.fgdc.gov/standards/projects/FGDC-standards-projects/riparian-mapping

**Comment by dwief-usgs** [comment]
@nmtarr I read through the report again.  Thanks for putting in a great level of detail.  Without delving into the details to much I think I have a general understanding of most of the methods and results coming out of the analysis, but below I outlined a few things that are unclear to me.

I could be missing something but wouldn't table 3 be more valuable if we also related it back to the SE values?  I feel like the SE is the target for the analysis but that table shows CONUS results only.

Also if you don't mind sharing a few of the specific grids used for the species overlays (referenced in Identifying floodplain forest species)  I think that will help me understand that overall process. Could you clarify if the SE analyses were still causing the large processing times or only when you were running at CONUS?  In other words, were the 1hr/sp overlays just for the SE?  I guess I am unclear for each species how many grids were being overlaid? Grids are 30m?  

Some of these might be easier to wait on until the call.  I'm sure you have a better sense of that.  Thanks!

**Comment by TWellman** [comment]
@nmtarr, I really appreciate the opportunity to peruse your report on the importance of Woody Wetlands. Rather than make direct suggestions prior to our upcoming meeting where objectives could evolve, I will only summarize a couple of my initial thoughts. Firstly, we need to think about the intended message of this work (i.e. what are you (we) trying to get across to the reader). Then we need to make those points abundantly clear in some type of effective message transfer format. I am not all that confident with the 50% cutoff analysis as an underpinning of this work. I recognize the threshold is somewhat arbitrary, as was stated, but it may be misleading as well. Conceptually, as a heuristic division you might divide the populations into thirds (0-.333, .334-666, .667-1.0). Species with about equal populations (.334-.666) of species between regions are probably largely insensitive to Woody Wetlands or Floodplain Forests, or can at least subsist in different regions. The other regions would be largely independent (0-.333) and moderately dependent ( .667-1.0), respectively. Using a 50% cutoff places the common middle coverage species into two opposing bins. The better approach, however, since heuristic divisions are often unsupported scientifically, may be to compare natural divisions in the data distribution. Moreover, the graphics are not all that effective in present form. Plot axes of common graphs use different scales, etc. making comparisons difficult. Some of the analysis results doesn't seem to be discussed (e.g. median vs mean). We might wish to add a few statistics to provide supporting evidence that richness in species is indeed greater between certain regions. Overall, the difference between SE and CONUS for all species is pretty close by eyeing the bar charts. The text in general needs less information in some areas and more in others. We might want to add an abstract as well, which could be injected into data files and ScienceBase items, etc. Readers will want the abstract description of the work readily at hand. All that said, there are several interesting project angles to explore, and I probably need to get up to speed on your exact method to nail things down a bit better. A good start .... thank you very much for sharing.

**Comment by nmtarr** [comment]
@McKerrow @TWellman @dwief-usgs Thanks for looking things over and for the comments.

@McKerrow 
I suppose it's worth exploring different ways of mapping the ecosystems of interest, but as we do that we need to keep in mind issues of thematic resolution, the extent of the area of interest, and the basis of links between our data and other data sets; were they based on maps, system names, or system descriptions?  The answers to those questions could would have cascading effects on the study's scope and questions.  I think it would be OK to change our definition and map of the ecosystems of interest, but any changes would need to either fit with current scientific question or some other clear and relevant one.  

@dwief-usgs 
We could add columns to Table 3 and maybe some readers would want to see that.  For me, the question of how well protected species are has a national extent, hence the CONUS results.  If a species were 80% protected in the SE but only 2% protected in CONUS, I would still consider it underrepresented on protected areas.  But again, I'm not sure there's any harm in adding SE columns and that information could likely be gathered from the analytic database.

I'll see if I can get a few example rasters onto Drive for you.  They have a CONUS extent.  Similar to the previous topic, they were not summed within the SE because I sought to assess the overlay at a national scale.  That ability to make a national assessment related to a regional feature is a strength of having national data that I like to emphasize.  One other thing to think about, SE extent analyses would take less time to process, but you add the step of clipping national data to the SE, which takes a long time too.

@TWellman 
I agree about shaping the analysis and format to be whatever is the most effective.  That's what I was getting at with reference to figuring out the "so what".  On the one hand, I think we would be trying to get across the identities of **species** and **locations**.  On the other hand, a list of species leaves me wanting to know more about the species, which is one area for development.  

Interesting comments about the cutoffs and potential use of bins.  We should definitely talk about what the options are for that.  I can see a few directions (some from your comments).
- Equal bins that produce "low", "medium", and "high" species lists and locations.
- Two bins from 50% cutoff
- Bins that are defined by natural breakpoints in the data.  Either one or multiple.  Note, that I was kind of thinking of the relevance of this when I reported the quantiles that species w/ over 50% fell in.
- Two bins, but the user gets to choose the cutoff using a slider bar in the map.

We can adjust the text and figures as needed.  It was all just a starting point and how we revise it will depend on what the outlet is, which is why it's good to discuss the analysis with you.  If it's going to be a journal article, then the Introduction and Discussion will need a lot more, and an Abstract.

@TWellman and @McKerrow, you both commented on the inconsistent axes of Figure 3.  Just so you know, I tried putting all curves on one graph but it wasn't helpful because the y-axis scale for CONUS is so much larger than for the SE and woody wetlands.   I think the value of showing the curves is their shape in relation to the x-axis.

**Comment by nmtarr** [comment]
@skybristol Should I move this discussion to another repository? gap_species? the woody wetland repository?  I suspect it could clog the feed here.

**Comment by dwief-usgs** [comment]
@nmtarr  thanks for sharing a few tifs.  I did pull a few into python and think from initial testing (which only accounts for reading in and adding 2 tifs so isnt complete) may show some significant improvements in performance.  I also noticed that all of the tifs you sent were different shapes which may be slowing down analyses in itself.  I am no expert with gridded data but don't a lot of the analytical advantages of using grids stem from the use of a consistent base grid?  Are these at different extents for storage issues or another reason?

Both tests imported tifs using GDAL into Python, then pushed the grids into numpy arrays.

Test1:  Clipped data to Florida just to make sure the process worked and was doing what I thought it was doing.  This loaded Florida tifs and added them in 0.103 minutes.  I verified the output shape (data.shape) and resulting values ( identify in ArcMap).   Everything looked good.

Test2: Clipped tifs in ArcMap to ensure same base grid (essentially clipped to CONUS) (this might take a little more poking around to do in Python itself but didn't want to invest unless we think its worth pursuing).   This process took 7.6 minutes to run.  A few times I ran this Jupyter returned memory issues so doing this outside of Jupyter will have to be explored to see if we still get similar potential issues. 

Nathan thanks again for the documentation with the code.  Very helpful!

If you think it would be helpful we can talk more about this before or during the call on Thur.

**Comment by nmtarr** [comment]
@dwief-usgs I'm glad to hear that you had some luck with that.  This issue of handling the addition of our habitat maps might make a good break-off topic/repository since it's pivotal to most analyses and will be critical to the goal of getting processing online.  

Regarding the extents, the versions I sent you are our _main_ versions, which we decided to keep as small as possible (extent) to reduce size constraints.  However, we realized the value in having versions with CONUS extents and "0" instead of "Null" values and we have local copies of those for all species.  They took weeks to produce, but we used functions [here](https://github.com/nmtarr/GAPAnalysis/blob/master/gapanalysis/data.py) that you could probably use too.  Analyses at the larger extent are slower as well.

Linda and I experimented with doing something similar to what you tried on an HPC using gdal and arcpy for comparison.  She needed 300GB of RAM to handle a 1GB tiff, but gdal was way faster when she got it running.  Here's a graph of results that I found in an email chain.
![image](https://user-images.githubusercontent.com/18449976/31671818-d87c5a2c-b328-11e7-804b-1b59f01676f6.png)

**Comment by dwief-usgs** [comment]
Thanks for sharing.  300GB RAM... wowzers!  Ya we should likely branch this off into another issue if you think its worth pursuing.  I think it would be a good exercise and something that we might be able to create a new function for and merge into future processes.

Does anyone know how to migrate some of the comments from here to another issue?  ??? @aulenbac @skybristol  @TWellman ???

**Comment by nmtarr** [comment]
This discussion has moved [here](https://github.com/nmtarr/SEWoodyWetlandAnalysis/issues).
